mbedtls/src/ssl/context.rs

```diff
@@ -500,6 +511,250 @@ impl HandshakeContext {
+            self.set_hostname(hostname)?;
+
+            let ptr = &mut *io as *mut _ as *mut c_void;
+            ssl_set_bio(
+                self.into(),
+                ptr,
+                Some(AsyncIoAdapter::<T>::call_send),
+                Some(AsyncIoAdapter::<T>::call_recv),
+                None,
+            );
+
+            self.io = Some(io);
+            self.inner.reset_handshake();
+        }
+
+        HandshakeFuture(self).await
+    }
+}
+
+#[cfg(all(feature = "std", feature = "async"))]
+impl<T: AsyncRead> AsyncRead for Context<AsyncIoAdapter<T>>
+where
+    AsyncIoAdapter<T>: IoCallbackUnsafe<AsyncStream>,
+{
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut TaskContext<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<IoResult<()>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(
+                IoErrorKind::Other,
+                "stream has been shutdown",
+            )));
+        }
+
+        self.io_mut()
+            .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?
+            .ecx
+            .set(cx);
+
+        let result = match unsafe {
+            ssl_read(
+                (&mut *self).into(),
+                buf.initialize_unfilled().as_mut_ptr(),
+                buf.initialize_unfilled().len(),
+            )
+            .into_result()
+        } {
+            Err(Error::SslPeerCloseNotify) => Poll::Ready(Ok(())),
+            Err(Error::SslWantRead) => Poll::Pending,
+            Err(e) => Poll::Ready(Err(crate::private::error_to_io_error(e))),
+            Ok(i) => {
+                buf.advance(i as usize);
+                Poll::Ready(Ok(()))
+            }
+        };
+
+        self.io_mut()
+            .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?
+            .ecx
+            .clear();
+
+        result
+    }
+}
+
+#[cfg(all(feature = "std", feature = "async"))]
+impl<T: AsyncWrite + Unpin> AsyncWrite for Context<AsyncIoAdapter<T>>
+where
+    AsyncIoAdapter<T>: IoCallbackUnsafe<AsyncStream>,
+{
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut TaskContext<'_>,
+        buf: &[u8],
+    ) -> Poll<IoResult<usize>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(
+                IoErrorKind::Other,
+                "stream has been shutdown",
+            )));
+        }
+
+        let buf = {
+            let io = self
+                .io_mut()
+                .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?;
+            io.ecx.set(cx);
+            io.write_tracker.adjust_buf(buf)
+        }?;
+
+        self.io_mut()
+            .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?
+            .ecx
+            .set(cx);
+
+        let result = match unsafe {
+            ssl_write((&mut *self).into(), buf.as_ptr(), buf.len()).into_result()
```

Comment:
Can't you implement this in terms of `self.send`? Same question for read.

Reply:
fixed